src/query/sql/src/planner/optimizer/optimizer.rs

```diff
@@ -223,56 +221,97 @@ pub fn optimize_query(opt_ctx: OptimizerContext, mut s_expr: SExpr) -> Result<SE
     // Run default rewrite rules
     s_expr = RecursiveOptimizer::new(&DEFAULT_REWRITE_RULES, &opt_ctx).run(&s_expr)?;
 
-    {
-        // Cost based optimization
-        let mut dphyp_optimized = false;
-        if opt_ctx.enable_dphyp && opt_ctx.enable_join_reorder {
-            let (dp_res, optimized) =
-                DPhpy::new(opt_ctx.table_ctx.clone(), opt_ctx.metadata.clone())
-                    .optimize(Arc::new(s_expr.clone()))?;
-            if optimized {
-                s_expr = (*dp_res).clone();
-                dphyp_optimized = true;
-            }
+    // Cost based optimization
+    let mut dphyp_optimized = false;
+    if opt_ctx.enable_dphyp && opt_ctx.enable_join_reorder {
+        let (dp_res, optimized) = DPhpy::new(opt_ctx.table_ctx.clone(), opt_ctx.metadata.clone())
+            .optimize(Arc::new(s_expr.clone()))?;
+        if optimized {
+            s_expr = (*dp_res).clone();
+            dphyp_optimized = true;
         }
-        let mut cascades = CascadesOptimizer::create(
-            opt_ctx.table_ctx.clone(),
-            opt_ctx.metadata.clone(),
-            dphyp_optimized,
-        )?;
-        s_expr = cascades.optimize(s_expr)?;
     }
+    let mut cascades = CascadesOptimizer::new(
+        opt_ctx.table_ctx.clone(),
+        opt_ctx.metadata.clone(),
+        dphyp_optimized,
+        enable_distributed_query,
+    )?;
+
+    // Cascades optimizer may fail due to timeout, fallback to heuristic optimizer in this case.
+    s_expr = match cascades.optimize(s_expr.clone()) {
+        Ok(mut s_expr) => {
+            s_expr =
+                RecursiveOptimizer::new(&[RuleID::EliminateEvalScalar], &opt_ctx).run(&s_expr)?;
```

Comment:
It may be better to put `CommuteJoin` in `RecursiveOptimizer`, because it can ensure that the number of rows in build side is smaller, while `CascadesOptimizer` may not be able to do this.

Reply:
The post optimization of `CommuteJoin` will effect the result of `Exchange` placement. I'll try to put it right after dphyp.

Reply:
Fixed.